ts2mls/js/src/main/scala/ts2mls/TSSourceFile.scala

```diff
@@ -0,0 +1,174 @@
+    })
+
+  private def getSubstitutionArguments[T <: TSAny](args: TSArray[T]): List[TSType] =
+    args.foldLeft(List[TSType]())((lst, arg) => arg match {
+      case token: TSTokenObject => lst :+ getObjectType(token.typeNode)
+      case tp: TSTypeObject => lst :+ getObjectType(tp)
+    })
+
+  private def getObjectType(obj: TSTypeObject): TSType =
+    if (obj.isEnumType) TSEnumType
+    else if (obj.isFunctionLike) getFunctionType(obj.symbol.declaration)
+    else if (obj.isTupleType) TSTupleType(getTupleElements(obj.typeArguments))
+    else if (obj.isUnionType) getStructuralType(obj.types, true)
+    else if (obj.isIntersectionType) getStructuralType(obj.types, false)
+    else if (obj.isArrayType) TSArrayType(getObjectType(obj.elementTypeOfArray))
+    else if (obj.isTypeParameterSubstitution) TSSubstitutionType(obj.symbol.escapedName, getSubstitutionArguments(obj.typeArguments))
+    else if (obj.isObject)
+      if (obj.isAnonymous) TSInterfaceType("", getAnonymousPropertiesType(obj.properties), List(), List())
+      else TSReferenceType(obj.symbol.fullName)
+    else if (obj.isTypeParameter) TSTypeParameter(obj.symbol.escapedName)
+    else TSPrimitiveType(obj.intrinsicName)
+
+  // get the type of a member in classes/named interfaces/anonymous interfaces
+  private def getMemberType(node: TSNodeObject): TSType = {
+    val res: TSType =
+      if (node.isFunctionLike) getFunctionType(node)
+      else getObjectType(node.`type`.typeNode)
+    if (node.symbol.isOptionalMember) TSUnionType(res, TSPrimitiveType("undefined"))
+    else res
+  }
+
+  private def getTypeParametes(node: TSNodeObject): List[TSTypeParameter] =
+    node.typeParameters.foldLeft(List[TSTypeParameter]())((lst, tp) =>
+      if (tp.constraint.isUndefined) lst :+ TSTypeParameter(tp.symbol.escapedName, None)
+      else lst :+ TSTypeParameter(tp.symbol.escapedName, Some(getObjectType(tp.constraint.typeNode)))
+    )
+
+  private def getFunctionType(node: TSNodeObject): TSFunctionType = {
+    val pList = node.parameters.foldLeft(List[TSType]())((lst, p) => lst :+ (
+      // in typescript, you can use `this` to explicitly specifies the callee
+      // but it never appears in the final javascript file
+      if (p.symbol.escapedName === "this") TSPrimitiveType("void")
+      else if (p.isOptionalParameter) TSUnionType(getObjectType(p.symbolType), TSPrimitiveType("undefined"))
+      else getObjectType(p.symbolType))
+    )
+    TSFunctionType(pList, getObjectType(node.returnType), getTypeParametes(node))
+  }
+
+  private def getStructuralType(types: TSTypeArray, isUnion: Boolean): TSStructuralType =
+    types.foldLeft[Option[TSType]](None)((prev, cur) => prev match {
+      case None => Some(getObjectType(cur))
+      case Some(p) =>
+        if (isUnion) Some(TSUnionType(p, getObjectType(cur))) else Some(TSIntersectionType(p, getObjectType(cur)))
+    }).get.asInstanceOf[TSStructuralType]
```

Comment:
`asInstanceOf` should be avoided. Why is it even important to return a `TSStructuralType` here?